src/parse.rs

```diff
@@ -870,34 +870,29 @@ where
             }
             match u16::from_str_radix(&s, 16) {
                 Ok(u) => {
-                    if (0xDC00..=0xDFFF).contains(&u) || (0xD800..=0xDB7F).contains(&u) {
-                        // Low/High Surrogates
+                    if (0xD800..=0xDBFF).contains(&u) {
+                        // Low Surrogates
                         if !self.try_consume_str("\\u") {
                             return Some(u as u32);
                         }
+                        orig_input = self.input.clone();
 
-                        let mut s = String::new();
-                        for _ in 0..4 {
-                            if let Some(c) = self.next().and_then(char::from_u32) {
+                        let result = (|| {
```

Comment:
I haven't seen this "immediately invoked function expression" pattern used before in Rust. Any reason not to just use a `{...}` block?

Reply:
Ah, this is just a poor's man try block 😆 it allows to use `?` inside a block without returning from the main function.

I had another version that used combinators instead, but it was a bit less readable than this imperative version.